### PR TITLE
Open sum types

### PR DESCRIPTION
**Pull request by rossabaker**
Represents every generated "sum type" as a newtype around `CI Text` rather than a sum type.

The definition of completeness lets us perform exhaustiveness checking on known values, while also gracefully handling new values that AWS has added since the we last regenerated the code.

For backward compatibility, an `Enum` and `Bounded` instance are kept, but these are disrecommended.

### CODE REVIEW

**Comment by rossabaker** [comment]
If anyone has a project that they'd like me to compile this on, let me know.  (Note: this repo is public, so Slack me if it's internal.)

If this works out well, I'll try to upstream it.

File: gen/src/Gen/Import.hs [hunk 1]
```diff
@@ -44,7 +44,8 @@ typeImports l = sort $
 
 sumImports :: Library -> [NS]
 sumImports l = sort $
-      "Network.AWS.Prelude"
+      "Data.CaseInsensitive"
```
Comment (rossabaker) [inline gen/src/Gen/Import.hs:47]: I would like to clean this up with a qualified import, which I currently think I'd have to embed directly in the .ede file.

**Comment by rossabaker** [comment]
There is still one compile failure on this in ec2, but something more troubling.  It now takes over 2.5 hours to build on this machine.

```
real	156m45.895s
user	1819m44.733s
sys	394m55.069s
```

**Comment by rossabaker** [comment]
For expediency of testing, this is rebuilt atop #15.

**Comment by rossabaker** [comment]
There are a few stubborn modules here, where a type constructor and data constructor share a name.  This doesn't fly once we turn the sum types' data constructors into pattern synonyms.

**Comment by andreyk0** [comment]
> There are a few stubborn modules here, where a type constructor and data constructor share a name. This doesn't fly once we turn the sum types' data constructors into pattern synonyms.

What's a small example of the generated problematic code?

**Comment by rossabaker** [comment]
This one I could work around, with a breaking change, by renaming the pattern if it's the same as the type constructor:

```hs
data Transferable =
  Transferable (CI Text)
  deriving (Eq, Ord, Read, Show, Data, Typeable, Generic)

pattern Transferable :: Transferable
pattern Transferable = Transferable "TRANSFERABLE"
```

vs.

```hs
data Transferable
  = DontKnow
  | Transferable
  | Untransferable
  deriving (Eq, Ord, Read, Show, Enum, Bounded, Data, Typeable, Generic)
```

There are also cases where a pattern name clashes another type constructor in the same module.  That one requires some additional bookkeeping.

In neither case have I figured out how to make it compatible, though it's rare, and we can probably live with it.

**Comment by rossabaker** [comment]
> There are also cases where a pattern name clashes another type constructor in the same module. That one requires some additional bookkeeping.

Which is also bad if a future clash develops.  We know that `Transferable` clashes immediately and could introduce that as `Transferable'`.  But if they later introduced an `DontKnow` type, it would ripple into this.

**Comment by andreyk0** [comment]
I think if you "prime" it it should be Ok:

```haskell
{-# LANGUAGE PatternSynonyms #-}

module Foo where

data Transferable = Transferable' String deriving (Eq, Ord, Show)

pattern Transferable :: Transferable
pattern Transferable = Transferable' "TRANSFERABLE"

pattern DontKnow :: Transferable
pattern DontKnow = Transferable' "IDK"

pattern Untransferable :: Transferable
pattern Untransferable = Transferable' "BOO"

-- and then

data DontKnow = DontKnow' String deriving (Eq, Ord, Show)

-- This should not be possible today, it would be same as having dup
-- constructor name in different types within the same module
-- pattern DontKnow :: DontKnow
-- pattern DontKnow = DontKnow' "not possible now"

pattern WhoKnows :: DontKnow
pattern WhoKnows = DontKnow' "should work"

```

**Comment by rossabaker** [comment]
Yeah, priming it was going to be my strategy if it conflicts.  We could prime everything, and be ready future conflicts, but then we've made a small breaking change into a massive breaking change.

You're right that a future `DontKnow` type couldn't have a `DontKnow` constructor today if `Transferable` already did.  But there was another example that's similar to having a `DontKnow` type without a `DontKnow` constructor.  So it's not enough to prime when we're rendering individual sums.  We have to know all the sum types and prime any pattern that matches any sum type name.

**Comment by rossabaker** [comment]
Okay, I'd been thinking of this backward.  If we prime the data constructor, instead of the patterns, life is good. Thanks to @andreyk0 for resetting my brain.

**Comment by rossabaker** [comment]
Some operations modules depend on the data constructors for the products.  These were exported by the extra `Product` modules, but not the `Types` module.  Reviving the `Product` module is the fastest way to get it to compile, though it would be nice to eliminate it

**Comment by rossabaker** [comment]
This compiles clean.  It would be still be nice to eliminate that internal module, but if we pause here, we can try it out and clean up later.

**Comment by andreyk0** [comment]
> This compiles clean. It would be still be nice to eliminate that internal module, but if we pause here, we can try it out and clean up later.

This is a great start! We can always clean up later but would be really nice to just try things out now.

**Comment by rossabaker** [comment]
brendanhay/amazonka#597

**Pull request merged**